src/wallet/sqlite.cpp

```diff
@@ -0,0 +1,632 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <wallet/sqlite.h>
+
+#include <chainparams.h>
+#include <crypto/common.h>
+#include <logging.h>
+#include <sync.h>
+#include <util/memory.h>
+#include <util/strencodings.h>
+#include <util/system.h>
+#include <util/translation.h>
+#include <wallet/db.h>
+
+#include <sqlite3.h>
+#include <stdint.h>
+
+static const char* DATABASE_FILENAME = "wallet.dat";
+static const int32_t WALLET_SCHEMA_VERSION = 0;
+
+static Mutex g_sqlite_mutex;
+static int g_sqlite_count GUARDED_BY(g_sqlite_mutex) = 0;
+
+static void ErrorLogCallback(void* arg, int code, const char* msg)
+{
+    // From sqlite3_config() documentation for the SQLITE_CONFIG_LOG option:
+    // "The void pointer that is the second argument to SQLITE_CONFIG_LOG is passed through as
+    // the first parameter to the application-defined logger function whenever that function is
+    // invoked."
+    // Assert that this is the case:
+    assert(arg == nullptr);
+    LogPrintf("SQLite Error. Code: %d. Message: %s\n", code, msg);
+}
+
+SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_path, bool mock) :
+    WalletDatabase(), m_mock(mock), m_dir_path(dir_path.string()), m_file_path(file_path.string())
+{
+    {
+        LOCK(g_sqlite_mutex);
+        LogPrintf("Using SQLite Version %s\n", SQLiteDatabaseVersion());
+        LogPrintf("Using wallet %s\n", m_dir_path);
+
+        if (++g_sqlite_count == 1) {
+            // Setup logging
+            int ret = sqlite3_config(SQLITE_CONFIG_LOG, ErrorLogCallback, nullptr);
+            if (ret != SQLITE_OK) {
+                throw std::runtime_error(strprintf("SQLiteDatabase: Failed to setup error log: %s\n", sqlite3_errstr(ret)));
+            }
+            // Force serialized threading mode
+            ret = sqlite3_config(SQLITE_CONFIG_SERIALIZED);
+            if (ret != SQLITE_OK) {
+                throw std::runtime_error(strprintf("SQLiteDatabase: Failed to configure serialized threading mode: %s\n", sqlite3_errstr(ret)));
+            }
+        }
+        int ret = sqlite3_initialize(); // This is a no-op if sqlite3 is already initialized
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to initialize SQLite: %s\n", sqlite3_errstr(ret)));
+        }
+    }
+
+    try {
+        Open(""); // mode is unused
+    } catch (const std::runtime_error& e) {
```

Comment:
```
msbuild /p:TrackFileAccess=false build_msvc\bitcoin.sln /m /v:q /nologo
C:\projects\bitcoin\src\wallet\sqlite.cpp(65,40): error C2220: the following warning is treated as an error [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
C:\projects\bitcoin\src\wallet\sqlite.cpp(65,40): warning C4101: 'e': unreferenced local variable [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
Command exited with code 1

Reply:
This could be worked around with the following patch:
```diff
--- a/build_msvc/common.init.vcxproj
+++ b/build_msvc/common.init.vcxproj
@@ -117,7 +117,7 @@
       <WarningLevel>Level3</WarningLevel>
       <PrecompiledHeader>NotUsing</PrecompiledHeader>
       <AdditionalOptions>/utf-8 /std:c++17 %(AdditionalOptions)</AdditionalOptions>
-      <DisableSpecificWarnings>4018;4221;4244;4267;4334;4715;4805;4834</DisableSpecificWarnings>
+      <DisableSpecificWarnings>4018;4101;4221;4244;4267;4334;4715;4805;4834</DisableSpecificWarnings>
       <TreatWarningAsError>true</TreatWarningAsError>
       <PreprocessorDefinitions>_SILENCE_CXX17_CODECVT_HEADER_DEPRECATION_WARNING;_SILENCE_CXX17_OLD_ALLOCATOR_MEMBERS_DEPRECATION_WARNING;ZMQ_STATIC;NOMINMAX;WIN32;HAVE_CONFIG_H;_CRT_SECURE_NO_WARNINGS;_SCL_SECURE_NO_WARNINGS;_CONSOLE;_WIN32_WINNT=0x0601;_WIN32_IE=0x0501;WIN32_LEAN_AND_MEAN;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <AdditionalIncludeDirectories>..\..\src;..\..\src\univalue\include;..\..\src\secp256k1\include;..\..\src\leveldb\include;..\..\src\leveldb\helpers\memenv;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
@sipsorcery ?

Reply:
What's the reason to not get rid of the unused variable `e`?

`catch (const std::runtime_error&) {
...
}`

Reply:
Removed `e`